Make Document::isTopDocument and Document::mainFrameDocument consistent

achristensen07 opened the pull request:
#### c20306a535f37402e25c1c7b6bb3bc4ca3485a68
<pre>
Make Document::isTopDocument and Document::mainFrameDocument consistent
<a href="https://bugs.webkit.org/show_bug.cgi?id=286097">https://bugs.webkit.org/show_bug.cgi?id=286097</a>
<a href="https://rdar.apple.com/143077999">rdar://143077999</a>

Reviewed by Brady Eidson.

They had different behavior when site isolation was on.
This makes them always return the main frame&apos;s document when site isolation is on.
This does not change behavior when site isolation is off.

http/tests/site-isolation/accessibility/hit-test-resolving-remote-frame.html
started failing after an earlier version of this change, but now it passes.

In order to keep videos working in iframes with site isolation on, I changed
MediaElementSession::playbackStateChangePermitted to not always forbid video
playing when the main frame document is null.  Instead, I check null and proceed.
There is still work to do here, but this will allow most behavior to work
as desired with regards to video and site isolation in the meantime.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::axObjectCache const):
(WebCore::Document::mainFrameDocument const):
(WebCore::Document::isTopDocument const): Deleted.
* Source/WebCore/dom/Document.h:
(WebCore::Document::isTopDocument const):
(WebCore::Document::isTopDocumentLegacy const): Deleted.

Canonical link: <a href="https://commits.webkit.org/289267@main">https://commits.webkit.org/289267@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5663161c3b22cc0f4b3d0af37e44216aae2d220a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/86174 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/5758 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/40524 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/91142 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/37031 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/88224 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/6003 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/13788 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/91142 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/37031 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/89179 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/4582 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78108 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/91142 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4401 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/32413 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/36124 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/74975 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/33275 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/92924 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/13411 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/9777 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/92924 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/13624 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73971 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/92924 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/18993 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/17391 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/6212 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13409 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/13433 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/18751 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/13202 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/16634 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/14988 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

beidson commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -3432,23 +3432,19 @@ AXObjectCache* Document::axObjectCache() const
-        mainFrameDocument->m_axObjectCache = makeUnique<AXObjectCache>(*mainFrameDocument);
+    ASSERT(rootFrameDocument.get() == this || !m_axObjectCache);
+    if (!rootFrameDocument->m_axObjectCache) {
+        rootFrameDocument->m_axObjectCache = makeUnique<AXObjectCache>(*rootFrameDocument);
```
I think this is an incorrect change in behavior.

The purpose of always storing the `AXObjectCache `on the main frame document is that so - for any given frame tree - there's exactly one `AXObjectCache`.

My reading of `rootFrame` code is that there can be multiple root frames in a given frame tree.
e.g. One LocalFrame in a site isolated process might end up with itself as the `rootFrame`, and a sibling `LocalFrame` could end up with **itself** as its `rootFrame`

I believe the proper fix for the `AXObjectCache` storage is that it lives on `Page` and is cleared when `Page::didChangeMainDocument()` is called.

achristensen07 committed:
Make Document::isTopDocument and Document::mainFrameDocument consistent
https://bugs.webkit.org/show_bug.cgi?id=286097
rdar://143077999

Reviewed by Brady Eidson.

They had different behavior when site isolation was on.
This makes them always return the main frame's document when site isolation is on.
This does not change behavior when site isolation is off.

http/tests/site-isolation/accessibility/hit-test-resolving-remote-frame.html
started failing after an earlier version of this change, but now it passes.

In order to keep videos working in iframes with site isolation on, I changed
MediaElementSession::playbackStateChangePermitted to not always forbid video
playing when the main frame document is null.  Instead, I check null and proceed.
There is still work to do here, but this will allow most behavior to work
as desired with regards to video and site isolation in the meantime.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::axObjectCache const):
(WebCore::Document::mainFrameDocument const):
(WebCore::Document::isTopDocument const): Deleted.
* Source/WebCore/dom/Document.h:
(WebCore::Document::isTopDocument const):
(WebCore::Document::isTopDocumentLegacy const): Deleted.

Canonical link: https://commits.webkit.org/289267@main
